### PR TITLE
Cart content subtotal

### PR DESCRIPTION
**Pull request by colliona**

### CODE REVIEW

File: cartModel.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
+module.exports = CartModel = function () {
```
Comment (kapusadam) [inline cartModel.js:1]: Please use native functions or lodash functions instead of the imperative approaches

File: index.js [hunk 2]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+app.use(bodyParser.json());
+
+app.post('/cart', function (req, res) {
+    var cartId = req.body.cartId;
+    var itemId = req.body.item._id;
+    var quantity = req.body.quantity;
+    var imageUrl = req.body.imageUrl;
+
+
+    if(cartId === 'undefined') {
```
Comment (kapusadam) [inline index.js]: How is cartId passed from the UI? In a nice world it shouldn't be string compared.

File: cartModel.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            cartToPutInto[i].quantity = quantity;
+            break;
+        }
+    }
+};
+
+CartModel.prototype.delete = function(cartId, itemId) {
+    var cartToDeleteFrom = this.cartById(cartId);
+
+    if(cartToDeleteFrom !== null) {
```
Comment (kapusadam) [inline cartModel.js]: Same here, use declerative approach

Comment (kapusadam) [reply cartModel.js]: Same for the put

File: index.js [hunk 4]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+
+function sendCartWithSubTotal(cartId, res) {
+
+    var userCart = cartModel.cartById(cartId);
+
+    var subTotal = 0;
+
+    if(userCart.length !== 0) {
+
+        var url = "http://localhost:5000/items?$filter=_id eq '" + userCart[0].itemId + "'";
```
Comment (kapusadam) [inline index.js]: domain with port should be used as a constant

File: index.js [hunk 5]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+
+        for (var i = 1; i < userCart.length; i++) {
```
Comment (kapusadam) [inline index.js]: Move this code to common function and make it reusable

File: index.js [hunk 6]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+            if (err) {
+                console.log(err);
+                res.send(500, "Server Error");
+                return;
+            }
+
+            var query = results[0].value;
+
+
+            for (var i = 0; i < userCart.length; i++) {
```
Comment (kapusadam) [inline index.js]: Lodash functions for help this check

File: index.js [hunk 7]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+
+    if(cartId === 'undefined'){
+        var uniqueId = _.uniqueId('cartId-');
+        cartModel.sessions.push({cart: [], cartId: uniqueId });
+        res.send({cart: [], cartId: uniqueId, subTotal: 0});
+    } else {
+
+        var userCart = cartModel.cartById(cartId);
+
+        if(userCart !== null){
```
Comment (kapusadam) [inline index.js]: Not big diffrence, but the code is readable if u check for positive things first

File: package.json [hunk 8]
```diff
@@ -15,6 +15,9 @@
     "mongodb": "^3.0.1",
     "request": "^2.85.0",
     "simple-odata-server": "1.0.0",
-    "simple-odata-server-mongodb": "^1.0.0"
+    "simple-odata-server-mongodb": "^1.0.0",
```
Comment (kapusadam) [inline package.json]: I know it was like this but remove the version range characters. It is not a good practice

File: cartModel.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        return sessionToReturn.cart;
+
+};
+
+CartModel.prototype.hasAlready = function(itemId, cartOfUser) {
+    var hasAlready = cartOfUser.find(function(item) {
+        return item.itemId === itemId;
+    });
+
+    if(hasAlready === undefined)
```
Comment (kapusadam) [inline cartModel.js]: just return !!hasAlready

File: cartModel.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        return true;
+};
+
+CartModel.prototype.addQuantity = function(itemId, cartOfUser, quantity) {
+
+    var foundItem = cartOfUser.find(function(item) {
+        return item.itemId === itemId;
+    });
+
+    if(foundItem !== undefined)
```
Comment (kapusadam) [inline cartModel.js]: if(foundItem)

Comment (kapusadam) [reply cartModel.js]: same for below

File: cartModel.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    }
+};
+
+CartModel.prototype.get = function(cartId) {
+
+    var sessionToReturn = this.sessions.find(function(session) {
+        return session.cartId === cartId;
+    });
+
+    if(sessionToReturn === undefined)
```
Comment (kapusadam) [inline cartModel.js]: use ternary operator here

Comment (kapusadam) [reply cartModel.js]: and return null instead of undefined

File: index.js (hunk 2 above)
Comment (kapusadam) [reply index.js]: Still 'undefined' is checked

File: index.js [hunk 12]
```diff
@@ -173,6 +179,135 @@ app.get("/continentFilter", function (req, res, next) {
...
+
+    if(cartModel.delete(cartId, itemId)) {
+        sendCartWithSubTotal(cartId, res);
+    } else {
+        res.status(404).send({error: 'Item not found'});
+    };
+});
+
+function makeUrlFromUserCart(userCart) {
+    var url = "http://localhost:" + PORT + "/items?$filter=_id eq '" + userCart[0].itemId + "'";
```
Comment (kapusadam) [inline index.js]: localhost can also go to constan

File: index.js [hunk 13]
```diff
@@ -173,10 +182,139 @@ app.get("/continentFilter", function (req, res, next) {
...
+            }
+
+            var query = results[0].value;
+
+            for (var i = 0; i < userCart.length; i++) {
+                var queriedItem = query.find(function(item) {
+                    return item._id === userCart[i].itemId;
+                });
+
+                resultArray.push({
```
Comment (laszloszegedi33) [inline index.js]: please check is the value of `queriedItem`

File: index.js [hunk 14]
```diff
@@ -173,10 +182,139 @@ app.get("/continentFilter", function (req, res, next) {
...
+            }
+
+            res.send({cart: resultArray, subTotal: subTotal});
+        });
+    } else {
+        res.send({cart: [], subTotal: 0});
+    }
+}
+
+app.get('/cart', function(req, res) {
```
Comment (laszloszegedi33) [inline index.js:296]: this should be better:
`var cartId = req.query.cartId;

if(cartId === undefined){
   var cartId = _.uniqueId('cartId-');
   cartModel.sessions.push({cart: [], cartId: cartId });
} else {
   var userCart = cartModel.cartById(cartId);

   if(userCart === undefined){
      cartModel.sessions.push({cartId: cartId, cart: []});
   }
 }
sendCartWithSubTotal(cartId, res);
`

**Review by laszloszegedi33 — APPROVED** [review]

**Review by kapusadam — APPROVED** [review]

**Pull request merged**